### PR TITLE
Allow blocking of some zombie specials

### PR DESCRIPTION
**Pull request by Saelfen**
#### Summary
Bugfixes "Allow blocking of some zombie specials"

#### Purpose of change
You can dodge certain zombie special attacks such as scratches, but they are impossible to block.

#### Describe the solution
Apply block to all locations that can be dodged in hardcoded monster specials, and have enough information to be blocked (body part and damage instance). Additionally allow block of JSONized monster attacks (though there are only four attacks at the moment that fall under this category, scratch being one of them).

Additionally, the individual checks for uncanny in hardcoded monster specials were integrated into dodge_check(), and allowed uncanny to trigger on_dodge() effects. 

#### Describe alternatives you've considered
Not make the change at all

generic call() used by JSONized special attacks use deal_melee_attack and hitspread to calculate a hit or miss. However, hardcoded special attacks use dodge_check instead. Questioned potentially reworking JSONized special attacks to use dodge_check as well

JSONize special attacks that could be JSONized. Figured this belongs in a separate PR instead.

#### Testing
Spawn zombie child which has the scratch special attack (JSONized) and see the scratch attack can be blocked.
Spawn tentacle dog which has the tentacle special attack (Hardcoded) and see the tentacle attack can be blocked.

### CODE REVIEW

**Comment by actual-nh** [comment]
@Venera3: Any comments?

**Comment by Venera3** [comment]
(I'm in the process of moving smash/reach attacks over to JSON, for context): I'd add a `blockable` bool to the monster attacks themselves, as well as potentially a block message (unless blocking prints its own message already). Otherwise bringing them more in line with normal melee attacks is a good thing balance-wise. The same is true for `dodge_check` vs `deal_melee_attack`, but I have to admit I'm a bit hazy on the difference.

**Comment by AccountAlias** [comment]
> (I'm in the process of moving smash/reach attacks over to JSON, for context):

From what I've seen on the "Impale" special attack which both has a JSON and hardcoded equivalent, the JSON overrides the hardcoded map so the JSONization of smash and reach to JSON would likely not collide with pre-existing hardcoded logic. They will just need to be deleted since they serve no more purpose (unless we would want to use the hardcoded values as a means of adding more effects on top of JSON specials instead)

> I'd add a `blockable` bool to the monster attacks themselves

Yes, blocking prints their own message already. I figured since there's no `dodgeable` flag, `blockable` wasn't necessary. Though, I can see that as the list grows, there could be a need for that. I would prefer this to be under a different PR though if possible.

> as well as potentially a block message (unless blocking prints its own message already).

Yes block messages are handled in block_hit()

> Otherwise bringing them more in line with normal melee attacks is a good thing balance-wise. The same is true for `dodge_check` vs `deal_melee_attack`, but I have to admit I'm a bit hazy on the difference. 

`dodge_check` is as follows while `deal_melee_attack` uses normal distribution to ascertain a hit roll. I believe `dodge_check` is some kind of mathematical approximation of the results of the normal distribution check deal_melee_attack ends up with -but I can't confirm for sure.
image

Most of the points addressed her are more of my surmisings and findings while adding block checks, they aren't actually handled in this PR

**Comment by Venera3** [comment]
Looking into it of the two `deal_melee_attack` seems to be the newer (and Creature-level) function, so I think moving away from the single-use `dodge_check` would actually be the way to go, with potentially integrating uncanny dodge rolls into `Character::get_dodge`. Out of scope for the moment, of course.

Unless you're planning on doing more stuff in this space I'd vote for including the blockability field in this one - it's relevant to the PR, and cleaner than hooking up any kind of automatic blockability calculation. Mitigating a smash that would send you flying with a cudgel feels a bit wonky, not that you can't do that with normal melee attacks.

**Comment by AccountAlias** [comment]
@Venera3 I hear you on that, dodgeable and blockable flags have been tested and committed. If not provided, special attacks are not blockable or dodgeable 😄

**Comment by Venera3** [comment]
Nice! I'd take that the other way, though - *most* attacks should be dodgeable/blockable, so it could be the default. The system already allows for overwriting these per-monster, anyway.

**Comment by AccountAlias** [comment]
They are now dodgeable and blockable by default

File: data/json/monster_special_attacks/monster_attacks.json [hunk 1]
```diff
@@ -63,7 +67,9 @@
     "hit_dmg_u": "The %1$s slams into you!",
     "hit_dmg_npc": "The %1$s slams into <npcname>!",
     "no_dmg_msg_u": "The %1$s tries to slam into you, but stumbles aside.",
-    "no_dmg_msg_npc": "The %1$s tries to slam into <npcname>, but fails to."
+    "no_dmg_msg_npc": "The %1$s tries to slam into <npcname>, but fails to.",
+    "dodgeable": true,
+    "blockable": true
```
Comment (I-am-Erk) [inline data/json/monster_special_attacks/monster_attacks.json:72]:
```suggestion
    "blockable": false
```
I think a slam is actually something you shouldn't be able to block, given that most of our blocks are parries.

Comment (actual-nh) [reply data/json/monster_special_attacks/monster_attacks.json:72]: What would stepping aside and tripping them count as?

Comment (AccountAlias) [reply data/json/monster_special_attacks/monster_attacks.json]: I agree, commited suggestion 😄

**Pull request merged**